GVRf/Framework/jni/shaders/material/custom_shader.cpp

```diff
@@ -29,91 +29,213 @@
+            };
+    addUniformKey(variable_name, key, f);
 }
 
-void CustomShader::addUniformVec2Key(std::string variable_name,
-        std::string key) {
-    int location = glGetUniformLocation(program_->id(), variable_name.c_str());
-    uniform_vec2_keys_[location] = key;
+void CustomShader::addUniformVec2Key(const std::string& variable_name,
+        const std::string& key) {
+    UniformVariableBind f =
+            [key] (Material& material, GLuint location) {
+                glm::vec2 v = material.getVec2(key);
+                glUniform2f(location, v.x, v.y);
+            };
+    addUniformKey(variable_name, key, f);
 }
 
-void CustomShader::addUniformVec3Key(std::string variable_name,
-        std::string key) {
-    int location = glGetUniformLocation(program_->id(), variable_name.c_str());
-    uniform_vec3_keys_[location] = key;
+void CustomShader::addUniformVec3Key(const std::string& variable_name,
+        const std::string& key) {
+    UniformVariableBind f =
+            [key] (Material& material, GLuint location) {
+                glm::vec3 v = material.getVec3(key);
+                glUniform3f(location, v.x, v.y, v.z);
+            };
+    addUniformKey(variable_name, key, f);
 }
 
-void CustomShader::addUniformVec4Key(std::string variable_name,
-        std::string key) {
-    int location = glGetUniformLocation(program_->id(), variable_name.c_str());
-    uniform_vec4_keys_[location] = key;
+void CustomShader::addUniformVec4Key(const std::string& variable_name,
+        const std::string& key) {
+    UniformVariableBind f =
+            [key] (Material& material, GLuint location) {
+                glm::vec4 v = material.getVec4(key);
+                glUniform4f(location, v.x, v.y, v.z, v.w);
+            };
+    addUniformKey(variable_name, key, f);
 }
 
-void CustomShader::addUniformMat4Key(std::string variable_name,
-        std::string key) {
-    int location = glGetUniformLocation(program_->id(), variable_name.c_str());
-    uniform_mat4_keys_[location] = key;
+void CustomShader::addUniformMat4Key(const std::string& variable_name,
+        const std::string& key) {
+    UniformVariableBind f =
+            [key] (Material& material, GLuint location) {
+                glm::mat4 m = material.getMat4(key);
+                glUniformMatrix4fv(location, 1, GL_FALSE, glm::value_ptr(m));
+            };
+    addUniformKey(variable_name, key, f);
 }
 
 void CustomShader::render(const ShaderUniformsPerObject& uniforms, RenderData* render_data,
         const std::vector<Light*> lightList, Material* material) {
-    for (auto it = texture_keys_.begin(); it != texture_keys_.end(); ++it) {
-        Texture* texture = material->getTextureNoError(it->second);
-        // If any texture is not ready, do not render the material at all
-        if (texture == NULL || !texture->isReady()) {
-             return;
+    initializeOnDemand();
+
+    {
+        std::lock_guard<std::mutex> lock(textureVariablesLock_);
```

Comment:
Copy on write definitely is good which I was about to mention. For this change, I think we can keep this as it is since it keeps the correctness, we may discuss later how to optimize those per-frame overheads including other places as separate changes.